lib/logdna/client.rb

```diff
@@ -1,164 +1,132 @@
-    end
-
-    def process_buffer(buffer_size)
-      if buffer_size > @actual_byte_limit
-        @buffer_over_limit = true
-        flush()
-        @buffer_over_limit = false
+      if @lock.try_lock
+        processed_message = process_message(msg, opts)
+        new_message_size = processed_message.to_s.bytesize
+        @buffer.push(processed_message)
+        @buffer_byte_size += new_message_size
+        @lock.unlock
+
+        flush if @flush_limit <= @buffer_byte_size
+        schedule_flush unless @flush_scheduled
       else
-        @task = create_flush_task
+        @side_message_lock.synchronize do
+          @side_messages.push(process_message(msg, opts))
+        end
       end
     end
 
-    # this should be running synchronously if @buffer_over_limit i.e. called from self.buffer
-    # else asynchronously through @task
-    def flush()
-      if defined? @@request and !@@request.nil?
-        request_messages = []
-        @lock.synchronize do
-          request_messages = @messages
-          @buffer.truncate(0)
-          @messages = []
-        end
-        return if request_messages.empty?
+    def send_request
+      @side_message_lock.synchronize do
+        @buffer.concat(@side_messages)
+        @side_messages.clear
+      end
 
-        real = {
-          e: 'ls',
-          ls: request_messages,
-        }.to_json
+      @request.body = {
+        e: "ls",
+        ls: @buffer
+      }.to_json
 
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+      handle_excpetion = lambda do |message|
+        puts message
+        @exception_flag = true
+        @side_message_lock.synchronize do
+          @side_messages.concat(@buffer)
         end
+      end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+      begin
+        @response = Net::HTTP.start(
+          @uri.hostname,
+          @uri.port,
+          use_ssl: @uri.scheme == "https"
+        ) do |http|
+          http.request(@request)
+        end
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          puts "The response is not successful "
         end
+        @exception_flag = false
+      rescue SocketError
+        handle_excpetion.call("Network connectivity issue")
+      rescue Errno::ECONNREFUSED => e
+        handle_excpetion.call("The server is down. #{e.message}")
+      rescue Timeout::Error => e
+        handle_excpetion.call("Timeout error occurred. #{e.message}")
+      ensure
+        @buffer.clear
       end
     end
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
```

Comment:
So correct me if I'm wrong but this seems weird to me - the @flush_scheduled is outside the lock, so it feels like you can get weird behavior. Isn't it only flush that needs to be locked? Can we just `synchronize` this whole section? Since we have `@flush_scheduled` if feels like that takes care of the lock waiting you're addressing. I'm also somewhat confused as to why we need locking on the side messages. Maybe explain to me.

Reply:
We need to lock side_messages bc some other threads can modify it and can have weird results.

Reply:
Regarding other points of your comment, I think you're probably right.

Reply:
Actually, if I change it in this way, side-messages lock will be redundant.

Reply:
No, I still need the lock for side-messages which has to be short. Because I am trying to optimize here the lock usage. If the client receives the messages but some other thread occupied the buffer, we can do the following: 
1. Wait for the lock release. Which can take a long time bc it might not be release until the request is done. We're holding the lock for the whole request process
2. Write those message in to the side-messages. Which are always protected by a short lock. (@side-messaged lock should be taken only for adding/removing data from side-messages). 

So choosing the number 2, I optimize the performance. Since we have to keep @lock for a long time, let's not block other threads from giving the messages away to the lib. and the lib will take care of them later.

Reply:
I don't think `schedule_flush`  should be inside any locks. But I'll move the `@flush_scheduled` flag switching into locks to prevent weird behavior.